File: app/src/main/java/org/schabi/newpipe/views/MarkableSeekBar.java
package org.schabi.newpipe.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

public class MarkableSeekBar extends FocusAwareSeekBar {
    public final List<SeekBarMarker> seekBarMarkers = new ArrayList<>();
    private final Paint markerPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final int markerMinWidthPx;
    private final int markerExtraHeightPx;

    public MarkableSeekBar(final Context context) {
        this(context, null);
    }

    public MarkableSeekBar(final Context context, final AttributeSet attrs) {
        this(context, attrs, com.google.android.material.R.attr.sliderStyle);
    }

    public MarkableSeekBar(final Context context,
                           final AttributeSet attrs,
                           final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        markerMinWidthPx = Math.max(1, Math.round(getResources().getDisplayMetrics().density * 2f));
        markerExtraHeightPx = Math.max(1, Math.round(getResources().getDisplayMetrics().density));
    }

    @Override
    protected void onDraw(@NonNull final Canvas canvas) {
        super.onDraw(canvas);
        drawMarkersInternal(canvas);
    }

    public void clearMarkers() {
        seekBarMarkers.clear();
        invalidate();
    }

    public void drawMarkers() {
        invalidate();
    }

    private void drawMarkersInternal(@NonNull final Canvas canvas) {
        if (seekBarMarkers.isEmpty()) {
            return;
        }

        final int trackTop = getCompatTrackTop() - markerExtraHeightPx;
        final int trackBottom = getCompatTrackBottom() + markerExtraHeightPx;
        final int rawTrackStart = getCompatTrackPositionForFraction(0f);
        final int rawTrackEnd = getCompatTrackPositionForFraction(1f);
        final int trackStart = Math.min(rawTrackStart, rawTrackEnd);
        final int trackEnd = Math.max(rawTrackStart, rawTrackEnd);

        for (final SeekBarMarker seekBarMarker : seekBarMarkers) {
            int start = getCompatTrackPositionForFraction((float) seekBarMarker.percentStart);
            int end = getCompatTrackPositionForFraction((float) seekBarMarker.percentEnd);
            if (start > end) {
                final int temp = start;
                start = end;
                end = temp;
            }

            if (end - start < markerMinWidthPx) {
                end = start + markerMinWidthPx;
            }

            start = Math.max(trackStart, start);
            end = Math.min(trackEnd, end);
            if (end <= start) {
                continue;
            }

            markerPaint.setColor(seekBarMarker.color);
            canvas.drawRoundRect(
                    start,
                    trackTop,
                    end,
                    trackBottom,
                    getTrackHeight() / 2f,
                    getTrackHeight() / 2f,
                    markerPaint
            );
        }
    }
}

File: app/src/main/java/org/schabi/newpipe/views/MarkableSeekBar.kt
package org.schabi.newpipe.views

import android.content.Context
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatSeekBar
import androidx.core.content.ContextCompat
import org.schabi.newpipe.R

open class MarkableSeekBar : AppCompatSeekBar {
    var seekBarMarkers: ArrayList<SeekBarMarker> = ArrayList()
    private var originalProgressDrawable: Drawable? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr)

    override fun setProgressDrawable(d: Drawable?) {
        super.setProgressDrawable(d)

        // stored for when we draw (and potentially re-draw) markers
        originalProgressDrawable = d
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)

        // re-draw markers since the progress bar may have a different width
        drawMarkers()
    }

    fun drawMarkers() {
        if (seekBarMarkers.isEmpty()) {
            return
        }

        // Markers are drawn like so:
        //
        //  - LayerDrawable (original drawable for the SeekBar)
        //    - GradientDrawable (background)
        //    - ScaleDrawable (secondaryProgress)
        //    - ScaleDrawable (progress)
        //    - LayerDrawable (we add our markers in a sub-LayerDrawable)
        //      - Drawable (marker)
        //      - Drawable (marker)
        //      - Drawable (marker)
        //      - etc...
        val width = measuredWidth - (paddingStart + paddingEnd)

        var layerDrawable = originalProgressDrawable as LayerDrawable?

        val markerDrawables = ArrayList<Drawable?>()
        markerDrawables.add(layerDrawable)

        for (seekBarMarker in seekBarMarkers) {
            val markerDrawable =
                ContextCompat.getDrawable(
                    context,
                    R.drawable.sponsor_block_scrubber
                )

            val colorFilter =
                PorterDuffColorFilter(seekBarMarker.color, PorterDuff.Mode.SRC_IN)

            checkNotNull(markerDrawable)
            markerDrawable.colorFilter = colorFilter

            markerDrawables.add(markerDrawable)
        }

        layerDrawable = LayerDrawable(markerDrawables.toTypedArray<Drawable?>())

        for (i in 1..<layerDrawable.numberOfLayers) {
            val seekBarMarker = seekBarMarkers[i - 1]
            val l = (width * seekBarMarker.percentStart).toInt()
            val r = (width * (1.0 - seekBarMarker.percentEnd)).toInt()

            layerDrawable.setLayerInset(i, l, 0, r, 0)
        }

        super.setProgressDrawable(layerDrawable)
    }

    fun clearMarkers() {
        seekBarMarkers.clear()
        super.setProgressDrawable(originalProgressDrawable)
    }
}
